### ISSUE TITLE
`CustomScrollView` applies overscroll effect on `SliverAppBar`

### ISSUE THREAD

**Issue by doiraduc** [flutter/flutter#104437]
Part of #104363.

When using the new variants SliverAppBar.medium and SliverAppBar.large in a CustomScrollView, scrolling in that CustomScrollView applies the overscroll effect on the SliverAppBar itself. This behavior should not be occuring.

**Expected result:**
- Scrolling through the CustomScrollView shows the overscroll effect only on the SliverList, not on the entire CustomScrollView.

**Actual result:**
- Scrolling through the CustomScrollView shows the overscroll effect on the SliverAppBar too, resulting in strange behavior.

https://user-images.githubusercontent.com/47219753/169891236-aeb82b09-fad6-427e-ad67-a0ff620d0708.mp4

<details>
<summary>Code sample:</summary>

```dart
import 'package:flutter/material.dart';

void main() {
  runApp(const TestApp());
}

class TestApp extends StatelessWidget {
  const TestApp({super.key});

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      theme: ThemeData.from(
        colorScheme: ColorScheme.fromSeed(seedColor: Colors.blue),
        useMaterial3: true,
      ),
      home: Scaffold(
        body: CustomScrollView(
          slivers: [
            SliverAppBar.large(
              title: const Text('Example App'),
            ),
            SliverList(
              delegate: SliverChildBuilderDelegate(
                (context, index) {
                  return ListTile(
                    title: Text('Item ${index.toString()}'),
                  );
                },
                childCount: 20,
              ),
            )
          ],
        ),
      ),
    );
  }
}
```
</details>

**However,** when the CustomScrollView scrollBehavior is set to CupertinoScrollBehavior, this issue does not occur.

https://user-images.githubusercontent.com/47219753/169891401-34de715b-caf7-4147-aec8-6483976f89b2.mp4

<details>
<summary>Code sample:</summary>

```dart
import 'package:flutter/material.dart';

void main() {
  runApp(const TestApp());
}

class TestApp extends StatelessWidget {
  const TestApp({super.key});

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      debugShowCheckedModeBanner: false,
      theme: ThemeData.from(
        colorScheme: ColorScheme.fromSeed(seedColor: Colors.blue),
        useMaterial3: true,
      ),
      home: Scaffold(
        body: CustomScrollView(
          scrollBehavior: const CupertinoScrollBehavior(),
          slivers: [
            SliverAppBar.large(
              title: const Text('Example App'),
            ),
            SliverList(
              delegate: SliverChildBuilderDelegate(
                (context, index) {
                  return ListTile(
                    title: Text('Item ${index.toString()}'),
                  );
                },
                childCount: 20,
              ),
            )
          ],
        ),
      ),
    );
  }
}
```
</details>

**Comment by Piinks** [comment]
I don't think this is an issue with the SliverAppBar, but a limitation of the stretching overscroll indicator.
The stretch is applied to the contents of the ScrollView, and the SliverAppBar is part of the scroll view. 

This is a similar issue that happens with the GlowingOverscrollIndicator, it paints on top of the SliverAppBar instead of below it. I have seen folks use the NestedScrollView in this case to place the SliverAppBar in the headerSliverBuilder, overscroll is directed to the inner scroll view (NestedScrollView.body), so the indicator is applied there, separate from the SliverAppBar. I have not tried it myself specifically with the Stretch indicator, but it should work to produce the effect you want. :)

FYI @darrenaustin

**Comment by iamcosmin** [comment]
Thank you!
It's strange that `CupertinoScrollBehavior` does not overscroll on the `SliverAppBar`, but meanwhile the new `MaterialScrollBehavior` overscroll effect has the same limitation as the old one. It's not the only issue I encountered with `StrechingOverscrollIndicator`, but that will be the subject of another issue. 
Though, if you intend to keep the behavior of the `StrechingOverscrollIndicator` as it is, I think it would be useful if the examples are using `NestedScrollView` with `SliverOverlapAbsorber` and `SliverOverlapInjector` instead of `CustomScrollView`, as the native counterpart of `SliverAppBar` does not show overscroll on the widget itself, thus the current behavior is not intended.
Let me know what you think.

**Comment by Piinks** [comment]
More sample code is always a great addition! I'll update this to reflect some doc updates on the feature and best cases for using it. :)

**Comment by kephren226** [comment]
> I don't think this is an issue with the SliverAppBar, but a limitation of the stretching overscroll indicator. The stretch is applied to the contents of the ScrollView, and the SliverAppBar is part of the scroll view.
> 
> This is a similar issue that happens with the GlowingOverscrollIndicator, it paints on top of the SliverAppBar instead of below it. I have seen folks use the NestedScrollView in this case to place the SliverAppBar in the headerSliverBuilder, overscroll is directed to the inner scroll view (NestedScrollView.body), so the indicator is applied there, separate from the SliverAppBar. I have not tried it myself specifically with the Stretch indicator, but it should work to produce the effect you want. :)
> 
> FYI @darrenaustin

Hello 

Please can you help me on using the sliverappbar medium 
When using `SliverAppBar.medium()` i got unknown constructor

**Comment by JonathanPeterCole** [comment]
I'm running into this as well. 

Using `NestedScrollView` gives the expected result if the scrollable content extends beyond the height of the screen, but if the content doesn't fill the height I run into https://github.com/flutter/flutter/issues/46028 where it scrolls behind the app bar. 

Using `SliverOverlapAbsorber` + `SliverOverlapInjector` resolves this, but the stretch effect then starts from the top of the screen rather than the bottom of the app bar, and the stretch effect still occurs when it doesn't need to scroll at all.
